### PR TITLE
fix: Authenticator issue where InitMachine useEffect runs every render

### PR DESCRIPTION
**Pull request by reesscot**
#### Description of changes
This PR fixes an issue where the `InitMachine` component is defined on each render. This caused the `useEffect` inside `InitMachine` to also run on each render, which finally caused the `children` of `InitMachine` to be unmounted and remounted. I also found that on initial render, the `useEffect` hook inside `InitMachine` was called many times. This PR fixes all of this by moving `InitMachine` outside of the `Authenticator` function. 

#### Issue #, if available

Fixes: https://github.com/aws-amplify/amplify-ui/issues/1451

#### Description of how you validated changes
Using the app from https://github.com/aws-amplify/amplify-ui/issues/1451 I validated the following:
* useEffect was only called a single time on initial render

* useEffect is not called when typing in the input field, and focus is not lost

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [x] Relevant documentation is changed or added (and PR referenced)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by wlee221 — APPROVED** [review]
Looks good to me, thanks for looking into it!

File: .changeset/nine-camels-cover.md [hunk 1]
```diff
@@ -0,0 +1,6 @@
+---
+"@aws-amplify/ui-react": patch
+---
+
+fix: Authenticator issue where InitMachine useEffect runs every render, causing `children` of 
+`Authenticator` to be unmounted and remounted on every render.
```
Comment (wlee221) [inline .changeset/nine-camels-cover.md:6]:
```suggestion
fix: Authenticator issue where InitMachine useEffect runs every render, causing `children` of `Authenticator` to be unmounted and remounted on every render.
```

NIT: You shouldn't need to manually add newlines, markdowns will handle it in changelogs.

**Review by jacoblogan — APPROVED** [review]
LGTM

File: packages/react/src/components/Authenticator/Authenticator.tsx [hunk 2]
```diff
@@ -20,6 +20,22 @@ export type AuthenticatorProps = AuthenticatorMachineOptions &
...
+  const { _send, route } = useAuthenticator();
+  React.useEffect(() => {
+    if (route === 'idle') {
+      _send({
+        type: 'INIT',
+        data: machineProps,
+      });
+    }
+    // Disabling exhaustive hooks rule here because we only want to run a single time
+    // eslint-disable-next-line react-hooks/exhaustive-deps
```
Comment (calebpollman) [inline packages/react/src/components/Authenticator/Authenticator.tsx:34]: Prefer passing `route` and `machineProps` to the deps array of the `useEffect`, the conditional on ln 27 will prevent the call to `_send` from executing over ignoring the **eslint** rule. 

Assuming there is potential of further calls to `_send` triggering when `route` updates to `'idle'` after the initialization call we can gate with a ref assigned variable (e.g. `hasInitialized`) to ensure that the call to `_send` only happens once. Thoughts?

File: packages/react/src/components/Authenticator/Authenticator.tsx [hunk 3]
```diff
@@ -20,6 +20,22 @@ export type AuthenticatorProps = AuthenticatorMachineOptions &
   RouterProps &
   ComponentsProviderProps;
 
+// Helper component that sends init event to the parent provider
+function InitMachine({ children, ...machineProps }) {
```
Comment (calebpollman) [inline packages/react/src/components/Authenticator/Authenticator.tsx:24]: Nit: replace `machineProps` with `data` to allow object shorthand notation on ln 30

**Review by calebpollman — COMMENTED** [review]
Thanks for taking care of this! Left some feedback and a nit

**Review by calebpollman — COMMENTED** [review]
LGTM 🚢

**Review by calebpollman — APPROVED** [review]

**Review by wlee221 — APPROVED** [review]

**Pull request merged**